### DISCUSSION TITLE
Where is v5?

### DISCUSSION THREAD

**Discussion by rpearce** [rpearce/react-medium-image-zoom#274] [General]
## Where is v5?

Hi, folx! Last year, I spent a couple of months trying all sorts of ideas out for v5 to solve various issues and make everything simpler and tinier, and I got burnt out and moved on to other things.

After my daughter was born last October, I briefly had some time where I revisited working on v5 and made good progress before the stress and lack of sleep made me feel awful for months.

I'm currently working on [another project, Ramda Guide](https://ramda.guide) (it's a small, free book), and I'm making progress every day. Once I get that out the door, I will circle back to `v5` as my top open source priority. It will be super simple to use, hyper-focused on images, _accessible_, and actually have quality zoomed images on iOS (ain't nobody doing that from what I can see). 

## What's that mean for v4?

That code has been abandoned for v5, but until v5 is released, I will make minor and patch updates to it if there are show-stopping issues.

* * *

Thanks for hanging in there with me!
Robert W. Pearce

**Comment by sunknudsen** [comment]
Hey Robert,

Thanks for the update and for creating image-zoom.

Having created image-zoom does not mean you are required to maintain the project forever (don't get me wrong, I am very thankful you have so far).

Life doesn't always make room for open source work... and one must not burn out. Congrats for your daughter btw!

I can empathize with you trying to not miss a beat... which is likely impossible. We're human after all.

❤️

**Reply by rpearce** [reply]
Thanks, @sunknudsen. Is it bad that I think about this every day? I'll get there, as I am a bit of a completionist, and there's pretty good usage of this, so I don't want to let peeople down.

Thank you for making it feel less of a burden.

**Reply by sunknudsen** [reply]
> Is it bad that I think about this every day?

No, it's human and totally normal.

> I'll get there, as I am a bit of a completionist, and there's pretty good usage of this, so I don't want to let peeople down.

If working on image-zoom makes you feel good, all good... If it has become a burden, well... not so good. Btw, thanks for sharing this thread publicly.

Too many people in open source burn out because of this kind of pressure. I wish I knew how to help, but I know more about computers than humans 🤓.

That said, thanks for all you have contributed (and might contribute in the future). In the meantime, you have my blessing to relieve my share of pressure (if any).

**Comment by rpearce** [comment]
Update: both my wife and my daughter have long-term illnesses now, so my time is limited when it comes to "what is the next open source thing I need to do?" We spent a week in the hospital with our daughter a couple of weeks ago and have been dealing with the medication she's on (that experience has gotten better recently). I'm not focused on this right now.

**Comment by rpearce** [comment]
Update: I'm looking at some small, incremental updates to first fix some accessibility issues like https://github.com/rpearce/image-zoom/issues/259 and the modal not being accessibly done at all.

We still have the home health issues and went back to the hospital for our daughter, but helping folx using this helps me feel a little better.

**Reply by sunknudsen** [reply]
Thanks for the update @rpearce.

> We still have the home health issues and went back to the hospital for our daughter, but helping folx using this helps me feel a little better.

Sending you guys some love. Heath is everything. 🤗

**Reply by rpearce** [reply]
Thanks a lot, @sunknudsen.

I have some potentially good ideas I'm working on here, but it takes time ⏲️ !

**Comment by rpearce** [comment]
Update: I've been working on the next major version, and it should keep the API _mostly_ the same, but the internals will work very differently, and it's going to be 100% accessible and should have the ability to include captions.

The going is slow, though, with work, my wife and daughter having long-term illnesses, etc.

**Reply by sunknudsen** [reply]
Thanks for the update Robert! 🤗

**Comment by rpearce** [comment]
Update: I've had a breaktrough with resolving some outstanding scaling & performance issues in a v5 rewrite branch I'm working on. This includes mobile iOS!

I've also got some fantastic accessibility updates for the modal in there, as well, and the ability to correctly work with gatsby / nextjs / whatever image components, too.

I've only got this for the `Uncontrolled` component at present, but this means I can now move on to things like:
* Providing a higher-res image to use (API is WIP on that... I'd love to use `srcset` or `sizes` on an `<img>` or pick them out of a `picture`, but that may prove to be too complex)
* Providing a caption (API is WIP)
* more...

Referenced issues that are getting attention here:
* #281 
* #259 
* #265 
* #264 
* #227 
* #213 
* #166 
* #164

**Comment by rpearce** [comment]
FYI that the [`v5-dev`](https://github.com/rpearce/react-medium-image-zoom/tree/v5-dev) branch is under heavy development and will solve a _ton_ of issues, makes this fully accessible, removes all dependencies, but also has a couple of notable breaking changes (some of which are up for discussion):

* Zooming images (`svg`, `img`, `img` inside `picture` or `figure`) is what this library is for. Zooming arbitrary things like `<div>`s won't be supported, _but_ if that `<div>` has a `background-image`, some magic will happen with zooming the full image. That latter point involves a lot of math and measuring!
* I'm dropping [`focus-options-polyfill`](https://www.npmjs.com/package/focus-options-polyfill), for this is something that is now implemented in Safari in some fashion at present, and if folx want/need to polyfill it for Safari's sake, they can do so
* _(up for debate)_ I'm planning on the JS only being outputted as ESModules
* _(up for debate)_ I'm planning on setting `"type": "module"` in `package.json`

If any of these build decisions above will affect you negatively, please respond here, and we can chat about it.

cc @sunknudsen if you want to chat more about it here

**Reply by rpearce** [reply]
_Update_

Things are going well, but I need to figure out the math for calculating how to smoothly zoom `<img />`s with [`object-fit`](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit) and `<div>`s with `background-image` and `background-size: contain | cover;`.

In `v4`, I'm merely taking the 4 corners of any object and scaling those to fill the X or Y axis to a certain point, and that's easy enough, but that doesn't allow for using an entire source image or loading a higher quality image on zoom.

For `v5`, I'll take the image source, put it in the modal at the same scale it's displayed at, then zoom that up and load a higher quality image and replace it once it's zoomed.

But I need to figure out this math, first.

**Reply by rpearce** [reply]
🎉 I've got the core math figured out for:

* ✅ Regular `<img />`, `<picture>`, and `<figure>`
* ✅ All of the above with any [`object-fit`](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit) _and_ no matter what the [`object-position`](https://developer.mozilla.org/en-US/docs/Web/CSS/object-position) is!
* 🚧 Doing the same with `<div>`s (or whatever) that use [`background-image`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-image) with [`background-size`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-size) and [`background-position`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-position) is a work in progress

There's still plenty of work to be done, but the first big hurdles have been solved.

* * *

To clarify a prior point: `v4` `<div>`s with `background-image` only scale what you visibly see, but `v5` will smoothly scale _the entire source image_, and this opens the door to replacing that image with a higher quality one.

**Reply by rpearce** [reply]
Update:

The math that I last reported on has made its way into functioning code for the ✅ items in my last post.

Honestly, it's a beautiful thing to see after all this work!

WIP things:

* 🚧 The offset from the browser window isn't being perfectly respected
* 🚧 `<div>` `background-image`s (should work almost exactly like some of the object fit ones, but we'll see...)
* 🚧 See how SVGs & `<img>`s with SVG `src` are handled

**Reply by rpearce** [reply]
Update: As of tonight, `<div>`s with `background-image`s are looking good! Here's the breakdown

* ✅ `background-size: cover;`
* ✅ `background-size: contain;`
* ✅ `background-size: auto;`
* 🚧 `background-size: <value1> <value2>;`

Once I get through this last math/measuring challenge above, I'll focus on providing a higher-res image to load on zoom.

And TBD on SVGs; I need to test those.

**Reply by rpearce** [reply]
Update: here's a short video of an image gallery using `<div>`s, `background-image`, `background-position`, and `background-size`. The example comes with various aspect ratios to try, as well.

https://user-images.githubusercontent.com/592876/176986921-fffe24e5-4a6d-4107-a8fe-324505b42788.mp4

**Reply by rpearce** [reply]
And here's a short video of an image gallery using `<img />`, all the `object-fit` options, and whatever you want `object-position` to be:

https://user-images.githubusercontent.com/592876/176987230-c426a7dc-7013-43fd-985c-fcff0b47ac1a.mp4

**Reply by rpearce** [reply]
Screencast updates I posted directly to issues:

* https://github.com/rpearce/react-medium-image-zoom/issues/166#issuecomment-1179716763
* https://github.com/rpearce/react-medium-image-zoom/issues/213#issuecomment-1179717981

I've also got the current build output down to three files whose _total_ is `5.3kb` (`index.js` is at `4.2kb` without client minification...I'll optimize further later):
1. `index.d.ts` (type definitions)
2. `index.js` (all the JS in a single ESModule)
3. `styles.css` (CSS styles)

**Reply by rpearce** [reply]
For all the zero people following this, here's a demo of a couple of SVGs that are now working:

https://user-images.githubusercontent.com/592876/179141752-41e5e07a-a635-46f5-ba44-d2df420c5655.mp4

SVG support involves a little bit more hackiness, but it's not that bad. There are still things to iron out, but at least we have a way of doing these things!

**Reply by sunknudsen** [reply]
Thanks for the updates… happy you are back at it. 🤓

**Comment by rpearce** [comment]
If anybody wants to play with the latest build, follow the directions here, https://github.com/rpearce/react-medium-image-zoom/issues/326#issuecomment-1193490590, and check out the README in [the `v5-dev` branch](https://github.com/rpearce/react-medium-image-zoom/tree/v5-dev).

Additionally, if anybody has opinions on `"type": "module"` and ESModule-only output, please weigh in here! https://github.com/rpearce/react-medium-image-zoom/discussions/328

**Comment by rpearce** [comment]
I've opened [a Draft PR](https://github.com/rpearce/react-medium-image-zoom/pull/329) to start moving this along. I need to write some basic, happy path functionality tests and see if anbody can beta test!

```
npm install --save react-medium-image-zoom@5.0.0-beta2.8
```

**Comment by rpearce** [comment]
This has been released as `5.0.0`. Any new conversations should go into a new discussion.

**Reply by lezan** [reply]
Hello @rpearce, I was using the `rc-10` and now I updated to `5.0.1`. The `useImageZoom`'s hook is no more a thing? Thanks!

**Reply by rpearce** [reply]
@lezan That hook never made it into `v5.0.0`. All of the work around that had to be abandoned, and the work picked back up beta testing with `5.0.0-beta2.x` released, as you can see here in NPM:

<img width="791" alt="imagen" src="https://user-images.githubusercontent.com/592876/184198511-3274c704-e5c0-4a29-84d5-30127f19cfc1.png">

I had previously thought that the hook was going to be the answer — to the point of almost releasing it! — but I ran into a brick wall with it a long while ago and had to rewrite everything.

I'm sorry if there was any confusion for you.

**Reply by lezan** [reply]
Hello @rpearce thanks as always for your quick and clear answer!
I was using the `rc-10` and it was great for my needs. I am using `useImageZoom` on a `<div>` to wrap `GatsbyImage` component from `gatsby-plugin-image`. I need to find an alternative way to make it working with the `Zoom` component.

**Reply by rpearce** [reply]
@lezan If you put `role="img"` on the div or `data-zoom`, it should work if you wrap it in the component.

**Reply by lezan** [reply]
Hello @rpearce thanks for your anwer. It does not work as excepted because `GatsbyImage` add a second `div` container and I am unable to forward the `role="img"` to the last one `div`.

**Reply by rpearce** [reply]
@lezan Can you tell me the steps to recreate this very simply, or even better: can you provide an example of how you're using it? A repo or online example?

If I have an example I can test, I can find a solution!

You may want to create a new issue, as well, with that info so others can track this.

**Reply by lezan** [reply]
Hello @rpearce sorry for the late reply, but I was out of town. I will provide you a repo online in a new issue! I just need a bit of time to wrap all the things.

**Reply by lezan** [reply]
Just for reference,
repo: https://github.com/lezan/zoom-test-gatsby
demo online: https://codesandbox.io/s/sharp-pike-fzr2ye?file=/src/pages/index.js

Going to create an issue soon.

**Reply by rpearce** [reply]
@lezan I've created an issue for you. I understand the issue, and I have a workaround for you in the issue: https://github.com/rpearce/react-medium-image-zoom/issues/347

**Reply by rpearce** [reply]
@lezan I've updated https://github.com/lezan/zoom-test-gatsby/pull/1, and if you could give the `5.0.2-beta.0` version a test, that would be helpful. It's looking like it's good to go, but I'd like some extra validation.
